Visibility

flodiebold opened the pull request:
This adds the infrastructure for handling visibility (for fields and methods, not in name resolution) in the HIR and code model, and as a first application hides struct fields from completions if they're not visible from the current module. (We might want to relax this again later, but I think it's ok for now?)

matklad commented on crates/ra_hir_def/src/db.rs:
```diff
@@ -90,6 +91,9 @@ pub trait DefDatabase: InternDatabase + AstDatabase {
     fn attrs(&self, def: AttrDefId) -> Attrs;
 
+    #[salsa::invoke(Visibility::visibility_query)]
+    fn visibility(&self, def: VisibilityDefId) -> Visibility;
```
Hm, I am not sure if we should really make this a query... Seems like it would be easier and more effective to just compute visibility directly in various `x_data` queries? Especially if we store an unresolved `ModPath` inside visibility.

matklad replied:
Hm, I wonder if that logic applies to generics as well... It seems reasonable to do all "data" lowering in once step, and then implement various traits on `XData` structs, rather than on IDs. (Take this with a grain of salt, as I am not exactly awake at the moment :)

flodiebold replied:
After working on integrating visibility in name resolution, I agree -- since visibility for items needs to be handled in the collector, this query would only really be used for methods and fields anyway.

matklad replied:
Hm, I now see that there's another problem here: we duplicate info between `raw` and `x_data` queries. 

For example, we extract `Name` out of `ast::StructDef` both in `RawItemsCollector::add_item` and in `struct_data_query`. Naturally, the two places do it differently: the former ignores anonymous structs, while the latter uses `Name::missing` (which I think is a dead code). 

But that probably is an unrelated problem. For know, I think it's ok to duplicate visbility as well between raw and data queries.

flodiebold replied:
I think in the case of visibility, we probably don't even need it in the item data. It makes more sense as a property of a specific (imported or defined) name as opposed to the item itself anyway, i.e. you shouldn't ask "is this item visible", but "is this path visible".

matklad commented on crates/ra_hir_def/src/visibility.rs:
```diff
@@ -89,6 +89,44 @@ impl Visibility {
+
+/// Visibility of an item, with the path resolved.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum ResolvedVisibility {
```
Not sure if it makes sense to distinguish between resolved and unresolved visibility. I guess producing a resolved one directly might be simpler? Especially if we special case `priv`, `pub` and `pub(crate)`, which don't need to resolve paths at all.

flodiebold replied:
I think the distinction is useful during name resolution, mirroring the raw items -> defs distinction :)

Btw, I would say that `pub(crate)` and `priv` need to resolve paths in the sense of determining the crate/the current module -- if we don't put that information into the resolved visibility, we'll need it later when checking the visibility, and then every caller would have to take care to pass the correct definition module.

flodiebold replied:
It might make sense to rename `Visibility` -> `RawVisibility` and `ResolvedVisibility` -> `Visibility` though :slightly_smiling_face:

matklad replied:
> need to resolve paths in the sense of determining the crate/the current module

Yes, we should store `ModuleId`s. My point is that resolving `pub(crate)` and `priv` is easy, in that you don't actually need to resolve an `hir::Path`, you need only the module structure.

flodiebold replied:
(I just did the rename using the LSP functionality, and the only reference it missed was the one in the `salsa::invoke` annotation -- maybe I shouldn't be impressed, but I am :slightly_smiling_face: )

matklad replied:
And note that the one in `salsa::invoke` can't handled correctly even in theory. It's a sad truth that a language with macros can't have precise rename because `macro_rules! fork_you { () => { $crate::unrenamable::path() } }` :(

matklad commented on crates/ra_hir/src/code_model.rs:
```diff
@@ -255,6 +255,15 @@ impl StructField {
+        let visibility = db.visibility(struct_field_id.into());
+        let parent_id: hir_def::VariantId = self.parent.into();
+        visibility.resolve(db, &parent_id.resolver(db))
+    }
```
One thing I don't like about this approach is that computing visibilites for all fields of the struct will repeatedly call `child_source` method on the struct itself, which seems wasteful. For `hir_def`, I've tried to mostly stick to design which avoids such needlessly duplicate queries, by returning an `ArenaMap` for all children instead (see child_sources or field_tys as examples). I am not sure if that's a good idea, but I'd like to try to stick with this approach, as it's more "data oriented" (we express shape of the data in queries, instead of hidhling it behind abstractions). 

But this maybe a moot point if we eagarly compute visibility and store it inside `XData`

flodiebold replied:
Yeah

matklad replied:
Ah, this is `code_model`! Yeah, for code_model repeated queries should not be a problem. As code_model is used mostly for rendering things to the user, it should not be on a hot path.

flodiebold commented:
@matklad I refactored how struct fields are handled, and also implemented visibility handling in name resolution. So this got quite a bit bigger, but I felt I needed to do the name resolution to validate the design :wink:

matklad commented on crates/ra_hir/src/code_model.rs:
```diff
@@ -1041,3 +1050,11 @@ impl<T: Into<AttrDef> + Copy> Docs for T {
+
+pub trait HasVisibility {
+    fn visibility(&self, db: &impl HirDatabase) -> Visibility;
+    fn visible_from(&self, db: &impl HirDatabase, module: Module) -> bool {
```
```suggestion
    fn is_visible_from(&self, db: &impl HirDatabase, module: Module) -> bool {
```

or `is_visible_at`

matklad commented on crates/ra_hir_def/src/item_scope.rs:
```diff
@@ -30,7 +33,7 @@ pub struct ItemScope {
     BuiltinType::ALL
         .iter()
-        .map(|(name, ty)| (name.clone(), PerNs::types(ty.clone().into())))
+        .map(|(name, ty)| (name.clone(), PerNs::types(ty.clone().into(), Visibility::Public)))
```
Hm, looks like builtins should have some other kind of visiblity, as you can't `use super::u32`, unless you also `use u32 as u32` in a parent module. But this probably doesn't matter outside of some extreme edge cases .

flodiebold replied:
I think that just needs to be handled in path resolution (i.e. we shouldn't consider the builtin scope when looking at a path with more than one segment? not sure if we currently do)

matklad replied:
Yeah, more generally, I think that treating builtins as just a separate scope was a mistake, we should rather special-case them in path resolution and in completions.

matklad commented:
bors r+
